dataproc/dataproc_quickstart/quickstart.go

```diff
@@ -0,0 +1,194 @@
+		ProjectId: projectID,
+		Region:    region,
+		Job: &dataprocpb.Job{
+			Placement: &dataprocpb.JobPlacement{
+				ClusterName: clusterName,
+			},
+			TypeJob: &dataprocpb.Job_PysparkJob{
+				PysparkJob: &dataprocpb.PySparkJob{
+					MainPythonFileUri: jobFilePath,
+				},
+			},
+		},
+	}
+
+	submitJobResp, err := jobClient.SubmitJob(ctx, submitJobReq)
+	if err != nil {
+		return fmt.Errorf("Error submitting job: %v", err)
+	}
+
+	id := submitJobResp.Reference.JobId
+
+	fmt.Fprintf(w, "Submitted job %q", id)
+
+	// These states all signify that a job has terminated, successfully or not
+	ts := map[dataprocpb.JobStatus_State]bool{
+		dataprocpb.JobStatus_ERROR:     true,
+		dataprocpb.JobStatus_CANCELLED: true,
+		dataprocpb.JobStatus_DONE:      true,
+	}
+
+	// We can create a timeout such that the job gets cancelled if not in a terminal state after a certain amount of time
+	timeout := 5 * time.Minute
+	start := time.Now().UTC()
+
+	state := submitJobResp.Status.State
+	for _, ok := ts[state]; !ok; _, ok = ts[state] {
+		// fmt.Fprintf(w, "-----\n")
+		// fmt.Fprintf(w, "State: %s", state)
+		// fmt.Fprintf(w, "out: %t, ok: %t", out, ok)
+		// fmt.Fprintf(w, "-----\n")
+		if time.Now().UTC().Sub(start) > timeout {
+			cancelReq := &dataprocpb.CancelJobRequest{ProjectId: projectID, Region: region, JobId: id}
+			jobClient.CancelJob(ctx, cancelReq)
+			return fmt.Errorf("Job %q timed out after threshold of %d seconds", id, int64(timeout.Seconds()))
+		}
+
+		time.Sleep(1 * time.Second)
+
+		getJobReq := &dataprocpb.GetJobRequest{
+			ProjectId: projectID,
+			Region:    region,
+			JobId:     id,
+		}
+		getJobResp, err := jobClient.GetJob(ctx, getJobReq)
+		if err != nil {
+			return fmt.Errorf("Error getting job %q with error: %v", id, err)
+		}
+
+		state = getJobResp.Status.State
+	}
+
+	// Cloud Dataproc job outget gets saved to a GCS bucket allocated to it.
+	getCReq := &dataprocpb.GetClusterRequest{
+		ProjectId:   projectID,
+		Region:      region,
+		ClusterName: clusterName,
+	}
+
+	resp, err := clusterClient.GetCluster(ctx, getCReq)
+	if err != nil {
+		return fmt.Errorf("Error getting cluster %q with error: %v", clusterName, err)
+	}
+
+	sc, err := storage.NewClient(ctx)
+	if err != nil {
+		return fmt.Errorf("Error creating storage client with error: %v", err)
+	}
+
+	obj := fmt.Sprintf("google-cloud-dataproc-metainfo/%s/jobs/%s/driveroutput.000000000", resp.ClusterUuid, id)
+	rc, err := sc.Bucket(resp.Config.ConfigBucket).Object(obj).NewReader(ctx)
+	if err != nil {
+		fmt.Fprintf(w, "UUID: "+resp.ClusterUuid)
+		fmt.Fprintf(w, "id: "+id)
+		return fmt.Errorf("Error reading Dataproc output: %v", err)
+	}
+
+	defer rc.Close()
+
+	body, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return fmt.Errorf("Could not read output from Dataproc Job %q", id)
+	}
+
+	fmt.Fprintf(w, "Job %q finished with state %s:\n%s", id, state, body)
+
+	// Delete the cluster once the job has terminated
+	dReq := &dataprocpb.DeleteClusterRequest{
```

Comment:
Should this be in a `defer` after the cluster is created to make sure the cluster gets cleaned up even if there is a return above? Need to be careful calling `log.Fatal`, though, since `defer`s don't run when you call `os.Exit`, which `log.Fatal` does.

Reply:
Is there a terminal `error` message that triggers `defer`?

Reply:
`panic(fmt.Sprintf("error: %v", err))`?

Reply:
Actually, I want to check with the Dataproc folks about this.